### PR TITLE
fix: Fix and test cron job for Invoice Due Marking

### LINKED ISSUE #6494: Cron job for marking event invoices due is wrong
PR https://github.com/fossasia/open-event-server/pull/6166 is wrong. The query fails in every case

https://github.com/fossasia/open-event-server/blob/bc45c1d267b924b556d36855f29c1dd96e901209/app/api/helpers/scheduled_jobs.py#L164-L174

This line is not valid SQLAlchemy syntax:

https://github.com/fossasia/open-event-server/blob/bc45c1d267b924b556d36855f29c1dd96e901209/app/api/helpers/scheduled_jobs.py#L169

`Neither 'InstrumentedAttribute' object nor 'Comparator' object associated with EventInvoice.event has an attribute 'ends_at'`

It should be `db.session.query(EventInvoice).join(Event).filter(EventInvoice.status == 'upcoming', Event.ends_at >= datetime.datetime.now()).all()`

Each run of this task is failing

Fix it and write tests to ensure it is working

### PR DESCRIPTION
**Pull request by mrsaicharan1**
Fixes #6494 

#### Short description of what this resolves:
Fixed event invoice marking error and added integration test for the same.

#### Changes proposed in this pull request:

- Used Multi-Value table update logic in accordance with the ORM norms
- Added integration test for the logic

### CODE REVIEW

File: tests/all/integration/api/helpers/test_scheduled_jobs.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    def test_event_invoices_mark_due(self):
+        """Method to test marking of event invoices as due"""
+
+        with app.test_request_context():
+            event_invoice_obj = EventInvoiceFactory()
+            event_obj = EventFactoryBasic()
+            event_obj.ends_at = datetime.date(2019, 7, 20)
+            event_invoice_obj.event = event_obj
+            save_to_db(event_obj)
+            save_to_db(event_invoice_obj)
```
Comment (iamareebjamal) [inline tests/all/integration/api/helpers/test_scheduled_jobs.py]: Can be changed to a single line

```python
event_invoice_obj = EventInvoiceFactory(event__ends_at=datetime.date(2019, 7, 20))
```

**Comment by mrsaicharan1** [comment]
Will update it asap.

File: tests/all/integration/api/helpers/test_scheduled_jobs.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import unittest
+from datetime import datetime
```
Comment (iamareebjamal) [inline tests/all/integration/api/helpers/test_scheduled_jobs.py]:
```suggestion
import datetime
```

**Comment by iamareebjamal** [comment]
https://stackoverflow.com/questions/42543223/how-do-you-express-a-multi-table-update-update-from-in-sqlalchemy-orm

**Comment by iamareebjamal** [comment]
I don't think you read the answer in SO question, you can't use update after using join.

Also, you can run the test locally before pushing, `nosetests tests/all/integration/api/helpers/test_scheduled_jobs.py`

File: app/api/helpers/scheduled_jobs.py [hunk 3]
```diff
@@ -164,12 +165,12 @@ def expire_pending_tickets():
...
     with app.app_context():
-        db.session.query(EventInvoice).\
+        db.session.query(EventInvoice).join(Event).\
                     filter(EventInvoice.status == 'upcoming',
-                           EventInvoice.event.ends_at >= datetime.datetime.now(),
+                           Event.ends_at >= datetime.datetime.now(),
                            (EventInvoice.created_at + datetime.timedelta(days=30) <=
                             datetime.datetime.now())).\
-                    update({'status': 'due'})
+                    update({'status': 'due'}, synchronize_session=False)
```
Comment (iamareebjamal) [inline app/api/helpers/scheduled_jobs.py]: `synchronize_session=False` has no use here

**Comment by mrsaicharan1** [comment]
@iamareebjamal I will test it out locally once before pushing.

File: app/api/helpers/scheduled_jobs.py (hunk 3 above)
Comment (mrsaicharan1) [reply app/api/helpers/scheduled_jobs.py]: Was just checking something else. Yeah.

File: app/api/helpers/scheduled_jobs.py [hunk 4]
```diff
@@ -164,14 +165,15 @@ def expire_pending_tickets():
...
     with app.app_context():
-        db.session.query(EventInvoice).\
-                    filter(EventInvoice.status == 'upcoming',
-                           EventInvoice.event.ends_at >= datetime.datetime.now(),
-                           (EventInvoice.created_at + datetime.timedelta(days=30) <=
-                            datetime.datetime.now())).\
-                    update({'status': 'due'})
-
-        db.session.commit()
+        event_invoices = db.session.query(EventInvoice).join(Event).\
```
Comment (iamareebjamal) [inline app/api/helpers/scheduled_jobs.py]: I don't know how else to say this, you cannot use join in the update query...

File: app/api/helpers/scheduled_jobs.py [hunk 5]
```diff
@@ -164,14 +165,15 @@ def expire_pending_tickets():
...
+                            filter(EventInvoice.status == 'upcoming',
+                            Event.ends_at >= datetime.datetime.now(),
+                            (EventInvoice.created_at + datetime.timedelta(days=30) <=
+                            datetime.datetime.now())).all()
+        for invoice in event_invoices:
+            obj = EventInvoice.query.filter_by(id=invoice.id).one()
+            obj.status = 'due'
+            save_to_db(obj)
```
Comment (iamareebjamal) [inline app/api/helpers/scheduled_jobs.py]: This is grossly inefficient. Please see the stackoverflow answer

**Comment by iamareebjamal** [comment]
Now that it's tested for an event that should be marked due. Check by creating another invoice that it should not be marked due to ensure that function is working correctly

**Comment by mrsaicharan1** [comment]
@iamareebjamal Okay, I think I swayed away from the StackOverflow Answer. So basically, without a join, the related attributes can still be accessed for an update and the strategy for updation was changed with respect to synchronisation. Alright

**Comment by mrsaicharan1** [comment]
I'll be adding for another test case where it shouldn't be marked as due.

File: tests/all/integration/api/helpers/test_scheduled_jobs.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            event_invoices_mark_due()
+            status = EventInvoice.query.get(event_invoice_id).status
+            self.assertEqual(status, "due")
+
+            event_invoice = EventInvoiceFactory(status="paid")
+            db.session.commit()
+            event_invoice_id = event_invoice.id
+            event_invoices_mark_due()
+            status = EventInvoice.query.get(event_invoice_id).status
+            self.assertNotEqual(status, "due")
```
Comment (mrsaicharan1) [inline tests/all/integration/api/helpers/test_scheduled_jobs.py]: Added the test case. @iamareebjamal

Comment (iamareebjamal) [reply tests/all/integration/api/helpers/test_scheduled_jobs.py]: There's one case missing for upcoming events with `(EventInvoice.created_at + datetime.timedelta(days=30) > datetime.datetime.now()`

Also, commit each one and run `event_invoices_mark_due()` once and verify the 3 invoice status

Comment (mrsaicharan1) [reply tests/all/integration/api/helpers/test_scheduled_jobs.py]: This test seems to fail. Created a future datetime object and passed it into the `created_at` attribute.

**Comment by iamareebjamal** [comment]
Seems like created_at check won't work this way. @mrsaicharan1 Can you tell me what does marking due means in terms of invoices?

**Comment by mrsaicharan1** [comment]
> Seems like created_at check won't work this way. @mrsaicharan1 Can you tell me what does marking due means in terms of invoices?

So the basic workflow here is that whenever an invoice is created, after 30 days from its creation, it would be marked as due [upcoming->due]

**Pull request merged**